lib/engine/game/g_18_al.rb

```diff
@@ -108,24 +112,52 @@ def stock_round
         ])
       end
 
-      def revenue_for(route, stops)
+      def route_hex_bonus(ability, stops)
+        stops.map do |stop|
+          next unless ability.hexes.include?(stop.hex.id)
+
+          { revenue: ability.amount, description: "Coal(#{stop.hex.name})" }
+        end.compact
+      end
+
+      def revenue_for(route, stops, routes: [])
         revenue = super
 
         abilities(route.corporation, :hexes_bonus) do |ability|
-          revenue += stops.sum { |stop| ability.hexes.include?(stop.hex.id) ? ability.amount : 0 }
+          revenue += route_hex_bonus(ability, stops).sum { |b| b[:revenue] }
         end
 
+        revenue += routes_revenue_bonuses(routes).sum { |b| b[:route] == route ? b[:revenue] : 0  }
+
         revenue
       end
 
-      def routes_revenue(routes)
-        total_revenue = super
-        route_bonuses.each do |type|
-          return total_revenue unless abilities(routes.first.corporation, type)
 
-          total_revenue += routes.map { |r| route_bonus(r, type) }.max
-        end if routes.any?
-        total_revenue
+      def routes_revenue_bonuses(routes)
+        return [] if routes.empty?
+
+        route_bonuses.map do |type, _description|
+          next unless abilities(routes.first.corporation, type)
+
+          possible_bonuses = routes.map { |r| route_bonus(r, type) }.compact
```

Comment:
What is the idea with the best bonus?

Reply:
This is almost the same code that was there before, but rather than applying the additional revenue immediately, route_bonus pairs the total bonus with the route that matched it. There could be multiple matches ( like ATL > BIR with two trains ), so the max call reduces it to applying to one route, the one where the bonus gives the best overall revenue.

Reply:
I wrote the original code ;-)

The problem here is that we have
- Route hex bonus
- Route bonus
- Routes bonus

And they can be unique or multiple.

18AL does have multiple route hex bonus, and two different unique routes bonuses.
Does the code allow for the case where two different (ATL>BRM,  MOB>NSH) routes bonuses are both triggered?

Reply:
I know you wrote it, that's why I requested your review. ;-)

Yeah, the naming is getting a little unwieldy. I'm in support of giving them more specific names. I'll do that.

"Does the code allow for the case where two different (ATL>BRM, MOB>NSH) routes bonuses are both triggered?"

Yeah, the variable `best_bonus` is really the best bonus for that bonus type ( atlanta-birmingham and mobile-nashville ). Note it is a map function, so we return results for both bonus types.

Reply:
I've got a local game where one company can just about run both routes, so I'll give it validation.

Reply:
<img width="856" alt="Screen Shot 2020-12-20 at 6 38 24 PM" src="https://user-images.githubusercontent.com/15675400/102730649-9a1c5900-42f2-11eb-98f8-230748ef5538.png">
<img width="554" alt="Screen Shot 2020-12-20 at 6 38 39 PM" src="https://user-images.githubusercontent.com/15675400/102730652-9b4d8600-42f2-11eb-9f9a-f70b46196119.png">